src/IotTelemetrySimulator/SenderBase.cs

```diff
@@ -0,0 +1,93 @@
+﻿using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace IotTelemetrySimulator
+{
+    public abstract class SenderBase<TMessage, TCommunicationException> : ISender 
+        where TCommunicationException : Exception
+    {
+        const int WaitTimeOnIotHubError = 5_000;
+
+        protected string deviceId;
+        protected RunnerConfiguration config;
+        protected Dictionary<string, object> variableValues;
+
+        public SenderBase(string deviceId, RunnerConfiguration config)
+        {
+            this.deviceId = deviceId;
+            this.config = config;
+        }
+
+        public abstract Task OpenAsync();
+
+        public async Task SendMessageAsync(RunnerStats stats, CancellationToken cancellationToken)
+        {
+            const int MaxAttempts = 3;
+            var msg = CreateMessage();
+
+            for (var attempt = 1; attempt <= MaxAttempts; ++attempt)
+            {
+                try
+                {
+                    await SendAsync(msg, cancellationToken);
+                    stats.IncrementMessageSent();
+                    break;
+                }
+                catch (TCommunicationException)
```

Comment:
Maybe if we define an virtual/abstract method called `bool IsTransientException(Exception)` we can remove the TCommunicationException. I think it is a bit easier to follow.

Reply:
Good idea.